jobmanager/AsyncWaiter.h

```diff
@@ -0,0 +1,195 @@
+#pragma once
+
+#include "jobmanager/JobQueue.h"
+#include "utils/Time.h"
+#include <vector>
+#include <functional>
+
+namespace jobmanager
+{
+
+class AsyncWaitTask
+{
+public:
+    virtual ~AsyncWaitTask() = default;
+
+    virtual bool checkCompletion() = 0;
+    virtual void onComplete() = 0;
+    virtual void onTimeout() = 0;
+};
+
+class AsyncWaiter
+{
+
+constexpr static uint32_t timerGroupId = 23963832U;
+
+public:
+struct AsyncEntry
+{
+    AsyncEntry(std::unique_ptr<AsyncWaitTask>&& task, uint64_t startTime, uint64_t endTime)
+        : task(std::move(task)),
+        startTime(startTime),
+        endTime(endTime)
+    { }
+
+    std::unique_ptr<AsyncWaitTask> task;
+    uint64_t startTime;
+    uint64_t endTime;
+};
+
+struct CheckJob : CountedJob
+{
+    CheckJob(AsyncWaiter& asyncWaiter)
+        : CountedJob(asyncWaiter._jobCounter),
+          _asyncWaiter(asyncWaiter)
+    { }
+
+    void run() final { _asyncWaiter.runCheck(); }
+
+    AsyncWaiter& _asyncWaiter;
+};
+
+struct ScheduleCheckJobTimer : CountedJob
+{
+    ScheduleCheckJobTimer(AsyncWaiter& asyncWaiter)
+        : CountedJob(asyncWaiter._jobCounter),
+          _asyncWaiter(asyncWaiter)
+    { }
+
+    void run() final { _asyncWaiter._jobQueue.template addJob<CheckJob>(_asyncWaiter); }
+
+    AsyncWaiter& _asyncWaiter;
+};
+
+struct AddTaskJob : CountedJob
+{
+    template<class...TaskArgs>
+    AddTaskJob(AsyncWaiter& asyncWaiter, uint64_t timeout, std::unique_ptr<AsyncWaitTask>&& task)
+        : CountedJob(asyncWaiter._jobCounter),
+         _asyncWaiter(asyncWaiter),
+          _task(std::move(task)),
+          _timeout(timeout)
+    { }
+
+    void run() final { _asyncWaiter.addNewTaskInternal(_timeout, std::move(_task)); }
+
+    AsyncWaiter& _asyncWaiter;
+    std::unique_ptr<AsyncWaitTask> _task;
+    uint64_t _timeout;
+};
+
+public:
+    explicit AsyncWaiter(JobManager& jobManager, size_t maxCheckInterval, size_t jobQueueSize = 128);
+
+    template<class Task, class... Args>
+    void emplaceTask(uint64_t timeout, Args&&...args);
+    template<class Task>
+    void addTask(uint64_t timeout, Task&& task);
+
+    bool hasTaskWaiting() const { return _jobCounter.load() != 0; }
+
+private:
+    void runCheck();
+    void scheduleNexCheck(uint64_t timeoutTs);
+    void addNewTaskInternal(uint64_t timeout, std::unique_ptr<AsyncWaitTask>&& task);
+    bool checkTaskCompletion(AsyncWaitTask& task);
+
+private:
+    JobQueue _jobQueue;
+    std::vector<AsyncEntry> _tasks;
```

Comment:
I am a bit sceptical to this data type. Workerthreads will have to do allocation. Could it be a more static array like StackArray. I know it is not ready for emplace and std:move, but perhaps it is an alternative.

Reply:
I thought a lot about this but I went with a simpler approach  because as is not used by EngineMixer and it creates a std::unique_ptr which needs a dynamic allocation anyway (although the allocation is on callee thread) the deallocation of unique_ptr will be on wt.

if we want to avoid dynamic allocations and have only a pure virtual type (AsyncWaitTask) we perhaps need to use memory pools here as well

Reply:
We could at least use reserve to have it ready for most likely scenarios.

Reply:
I reserve 16 by default. I don't expect having much pending task in simultaneity. Each mixer has their own queue